Refactoring of multiple components: MNIST MLBox, SSH and Singularity runners

sergey-serebryakov opened the pull request:
## Summary

### Runners
- New Docker, Singularity and SSH runners.
- Docker runner was refactored to provide the same CLI as other runners.
- SSH runner supports Docker and Singularity runners on remote hosts.

### Documentation
- Docker, Singularity and SSH runners.
- MNIST MLBox example.
- Hello World MLBox example.

### MLBoxes
- Hello World was refactored to support all runners.

### Issues
This PR implements #49, #61 and #64.

xyhuang commented:
here are some minor comments:
- shall we rename various occurrences of mlbox to mlcommons-box
- maybe the directory `mnist/platform` can use plural form instead (`mnist/platforms`), so that it can be consistent with how existing directories are named (e.g. `mnist/tasks`)

swiftdiaries commented on examples/hello_world/build/hello_world.py:
```diff
@@ -0,0 +1,35 @@
+from datetime import datetime
+
+
+app = typer.Typer()
```
I am so sorry but it looks like `typer` is still too early :(
With [click](https://click.palletsprojects.com/en/7.x/), [as shown here in `mlcommons_box`](https://github.com/mlperf/mlbox/blob/master/mlcommons_box/mlcommons_box/main.py#L13-L42) we can add tests with the in-built test framework that click provides.
[Testing reference for `mlcommons_box`](https://github.com/mlperf/mlbox/blob/master/mlcommons_box/mlcommons_box/tests/test_mlcommons_box_cli.py)

swiftdiaries commented on docs/runners/ssh-runner.md:
```diff
@@ -0,0 +1,57 @@
+## Run command
+During the run phase, the SSH runner performs the following steps:  
+1. It uses `ssh` to run standard `run` command on a remote host.  
+2. It uses `rsync` to synchronize back the content of the `{MLBOX_ROOT}/workspace` directory.   
```
💯  on the detailed docs

swiftdiaries commented on mkdocs.yml:
```diff
@@ -5,8 +5,14 @@ docs_dir: 'docs/'
+    - Runners: runners/index.md
+    - Docker Runner: runners/docker-runner.md
+    - Singularity Runner: runners/singularity-runner.md
+    - SSH Runner: runners/ssh-runner.md
```
💯

swiftdiaries commented on runners/mlbox_docker_run/mlbox_docker_run/__main__.py:
```diff
@@ -0,0 +1,35 @@
+from mlbox_docker_run.docker_run import DockerRun
+
+
+app = typer.Typer()
```
Same comment as above with the usage of Click over Typer. Again, so sorry for the inconvenience

swiftdiaries commented on runners/mlbox_docker_run/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+typer
+mlspeclib
```
Do we want to version the libraries?

swiftdiaries commented on runners/mlbox_singularity_run/mlbox_singularity_run/__main__.py:
```diff
@@ -0,0 +1,36 @@
+
+
+metadata.MLSchemaRegistrar.register()
+app = typer.Typer()
```
Same comment as above with respect to the usage Click over Typer. Again, sorry :(

swiftdiaries commented on runners/mlbox_singularity_run/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+typer
+mlspeclib
```
Same comment as above with respect to versioning of libraries

swiftdiaries commented:
+1 on using `entrypoints` to make the runner CLI a standalone implementation

swiftdiaries commented:
Also, could you check the test failure?
And if possible add basic tests and CI for the runners? (We can file an issue and do this in a follow-up PR too)

sub-mod commented:
I think the PR needs a rebase before merging.

sergey-serebryakov commented:
@xyhuang @sub-mod @swiftdiaries @bitfort 
Thanks for review! I think I addressed some of them. I also have several questions about others.

> if we want to improve the UX a little bit, we may add an entry_points property in the runner's setup scripts, so that user can run the runner by directly calling mlbox_docker_run instead of python mlbox_docker_run. refer to the setuptools doc or this stackoverflow answer

[__not fixed__] MLCommons-Box runners already provide `__main__.py` files that define entry points. In this [stackoverflow answer](https://stackoverflow.com/a/45143896/575749) they say that setup.py is not required to provide entrypoints in this case. Or do we still need to define entrypoint because that's considered to be a good practice?

> shall we rename various occurrences of mlbox to mlcommons-box

[__not fixed__] I looked at various sources where MLBox is mentioned and wanted to get your inputs before I update files. 
1. How do we refer to MLBox library and MLBox instances in documentation files?
    - `MLCommons-Box` for the library?
    - `MLCommons-Box runners` for runners in general?
    - `MLCommons-Box RUNNER` for concrete runners, such as `MLCommons-Box Docker Runner`?
    - `MLCommons-Box boxes` for packaged workloads in general?
    - `BOX-NAME MLCommons-Box ` for concrete instances, such as `Hello World MLCommons-Box`, `MNIST MLCommons-Box`, `Transformer MLCommons-Box`?

Using these full qualified names makes documentation pages look not cool. Assuming the above notation has been introduced, maybe we can just use `the library`, `runners` and `Docker Runner` for the first three items. I am not sure about good short terms for the last two: `boxes`, `MNIST box`?
2. Do we need to rename directories for the reference runners?
   - `mlbox_docker_run` -> `mlcommons_box_docker_run`
   - `mlbox_singularity_run` -> `mlcommons_box_singularity_run`
   - `mlbox_ssh_run` -> `mlcommons_box_ssh_run`

> maybe the directory mnist/platform can use plural form instead (mnist/platforms), so that it can be consistent with how existing directories are named (e.g. mnist/tasks)

[__fixed__]. What about `run` directory?

> Move from typer to click

[__fixed__].

> Do we want to version the libraries in the requirements files?

[__fixed__].

> Also, could you check the test failure?

[__fixed__]. I think it was due to the wrong version of mlspeclib library (last week).

> And if possible add basic tests and CI for the runners? (We can file an issue and do this in a follow-up PR too)

[__not fixed yet__]. 

> I think the PR needs a rebase before merging.

Is this to keep the commit history? I do not know much about it, what about squash merge since I do not think intermediate commits are useful from historical perspective? 


### Other questions
The `setup.py` files are empty now for runners. Can I just reuse mlcommons-box's implementation with some minor changes?

swiftdiaries commented:
Thanks for the changes @sergey-serebryakov !
> Or do we still need to define entrypoint because that's considered to be a good practice?

I think we want some conformance across python packages in the repo. So if we decide on one and stick with it throughout, it's good practice for readability and for onboarding developers.

And secondly, I think we want all runners to be independently publishable to PyPi. So each runner is it's own standalone CLI/package.

> Using these fully qualified names makes documentation pages look not cool.

I agree. We can codify some notations and denote it at the start of the docs and use it recurrently throughout the repo. I don't really have an opinion on the choices. I think the ones you picked are fine. Regarding the example mlboxes, we can call them `boxes` in plural and `<use-case> box` similar to how you proposed can be an option.

>  Do we need to rename directories for the reference runners?

Yes, I think we can rename them so that whatever the PR team comes up with we can uniformly search-and-replace. If we have divergent names, we might have some potential toil when we have to rename.

>  Can I just reuse mlcommons-box's implementation with some minor changes?

Yes, we'll automatically ensure commonality across the repo and developers adding new runners will also tend to follow established code practices. I think we want repeatable patterns in the common library and the runners.

> The __main__.py files of all three runners look exactly the same. Two differences - runner and metadata classes. I guess it can be refactored somewhow.

This is probably a minor nit imho. If you feel strongly about it, we can file an issue to better address it in a later sprint.

sergey-serebryakov commented:
Hi! I addressed 3 comments:
- Renamed various occurrences of `mlbox` to `mlcommons_box`. This includes documentation, code comments and directory structure (that resulted in different package names for runners - `mlcommons_box_docker_run` etc.).
- Added implementation for setup.py files including entrypoint specification. This is  basically a copy of the mlcommons_box's setup.py
- Added primitive CLI unit tests and CI workflow steps for runners. Same as current mlcommons_box implementation.
- Removed existing content from runners' README files, added references to documentation pages.


## TODO
1. Rebase.
2. I need your advice. Runners depend on mlcommons_box package. This includes, for instance, the metadata parser.
   - Is it a good idea to have this package dependency (runner -> mlcommons_box)?
   - What will runners' `requirements.txt` files contain - something like `mlcommons_box==0.1`?

sergey-serebryakov commented:
Rebase done. Remaining question - runners' dependency on MLCommons-Box library.

swiftdiaries commented:
Hey @sergey-serebryakov 
Sorry for the late review.
Thank you so much for the patience with this PR 💯 

>  need your advice. Runners depend on mlcommons_box package. This includes, for instance, the metadata parser.
    Is it a good idea to have this package dependency (runner -> mlcommons_box)?
    What will runners' requirements.txt files contain - something like mlcommons_box==0.1?

I think we want that dependency with the way we've structured the common libraries and schema.
Yes, the `requirements.txt` will have the same versioning as the mlcommons_box. If you're making changes with both the packages, the expectation is to change version numbers on both ends

sub-mod commented:
Merging based on everyone's approval .
